### PR TITLE
Report a dummy resource if sensitive is enabled

### PR DESCRIPTION
**Pull request by afiune**
### Description

Before this commit we were sending sensitive resource information to
Reporting and therefor you were able to see the sensitive data on the
Run History in the Chef Manage Console.

This commit is fixing this problem and now it is transforming any
sensitive resource to display the word `*sensitive*` instead of the
actual data. As we are inserting this values directly to the data base,
it means that this change will cascade all the way to the front-end.

Catch: Old data that was already reported and is displaying sensitive
data will contineu to be displayed. Apologize.

Signed-off-by: Salim Afiune <afiune@chef.io>

### Issues Resolved

COOL-642/ZD 12936 - Chef Manage Run History compromises sensitive data

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] RELEASE\_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by afiune** [comment]
Here some screenshots of the actual result on the Chef Manage Console:

screen shot 2016-12-20 at 5 16 29 pm

**Clic on the** 👁 

screen shot 2016-12-20 at 5 16 36 pm

**Comment by thommay** [comment]
I don't think we should transform the name property, since that just makes it hard for users to see what's going on. 
I understand that that could result in a leak of sensitive information for execute resources, but I'd prefer to see us raise an error in the client run if the resource is set as sensitive and the `command` parameter is unset, since we're being put in a position where we can't help but leak information. 
@chef/client-core thoughts?

**Comment by tduffield** [comment]
I kind of agree with Thom here. The initial card was focused around the execute command being shown: is there a solution that is specific to that case rather than the more general solution we have here?

**Comment by lamont-granquist** [comment]
it sort of feels like the name property should be left unchanged, while any other property should be omitted if the resource is marked sensitive, and we shouldn't be doing one-offs for file, execute, etc.

**Comment by coderanger** [comment]
This looks like it is mutating the resource objects in place. Am I missing a dup?

**Comment by coderanger** [comment]
Also should this match the behavior and formatting for the existing display code, not just re: name but in general?

**Comment by afiune** [comment]
@thommay @tduffield @lamont-granquist @coderanger I really appreciate the feedback!!! I am very open to suggestions around the implementation since this is the first time I am modifying Reporting (ish) - Would it make sense to create a zoom session to discuss? If not I'll ping you on Slack. 👍

**Comment by tduffield** [comment]
Kay. Which means we'll want to do a resource.dup, modify :allthethingss:, and then return that dup'd resource :)

**Comment by coderanger** [comment]
That's one option, another is to do the redaction when building the hash form of the data rather than where it is in the this patch. That seems less error prone as there would be no mutation of existing objects.

**Comment by tduffield** [comment]
We should make sure to update the README with this.

**Comment by afiune** [comment]
@tduffield You mean the RELEASE_NOTES. - Yeah good idea. On it! 👍

**Comment by afiune** [comment]
I had to rebase... But I think this is ready!! 



cc/ @chef/maintainers

**Comment by afiune** [comment]
@coderanger What do you think about this implementation?

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -6,10 +6,21 @@ _This file holds "in progress" release notes for the current release under devel
...
 
+## Highlighted bug fixes for this release:
+
+- Fixed exposure of sensitive data of resources marked as sensitive inside Reporting. Before you
+  were able to see the sensitive data on the Run History tab in the Chef Manage Console. Now we
+  are sending a new blank resource if the resource is marked as sensitive, this way we will not
+  compromise any sensitive data.
+
+  _Note: Old data that was already sent to Reporting marked as sensitive will continue to be
+  displayed. Apologize._
```
Comment (coderanger) [inline RELEASE_NOTES.md]: This should be fixed to at least be grammatically correct.

File: lib/chef/resource_reporter.rb [hunk 2]
```diff
@@ -198,6 +198,17 @@ def resource_failed(new_resource, action, exception)
...
         @pending_update.finish
+
+        # Verify if the resource has sensitive data
+        # and create a new blank resource with only
+        # the name so we can report it back without
+        # sensitive data
+        if @pending_update.new_resource.sensitive
+          klass = @pending_update.new_resource.class
+          resource_name = @pending_update.new_resource.name
+          @pending_update.new_resource = klass.new(resource_name)
```
Comment (coderanger) [inline lib/chef/resource_reporter.rb]: This leaves no way to know if a resource was actually blank or sensitive:

```ruby
package 'foo'

# vs

package 'foo' do
  sensitive true
end
```

I don't really need to worry much about y'alls commercial features so I suppose it's up to you if you think that's a problem. I will also point out that we already have a method for fixing this, `Chef::Resource#state_for_resource_reporter`.

**Comment by coderanger** [comment]
Also I don't think this handles sensitive props? I guess that would actually be a problem if that's something you intend to support.

**Comment by afiune** [comment]
@coderanger I just fixed the typo. Thank you for the great review. I will merge this as soon as test pass.

**Pull request merged**